bam_sort.c

```diff
@@ -2740,12 +2739,20 @@ int bam_sort_core_ext(SamOrder sam_order, char* sort_tag, int minimiser_kmer,
 
         if (k == max_k) {
             bam1_tag *new_buf;
+			template_coordinate_key_t *new_keys;
             max_k = max_k? max_k<<1 : 0x10000;
             if ((new_buf = realloc(buf, max_k * sizeof(bam1_tag))) == NULL) {
                 print_error("sort", "couldn't allocate memory for buf");
                 goto err;
             }
             buf = new_buf;
+			if (sam_order == TemplateCoordinate) {
+				if ((new_keys = realloc(keys, max_k * sizeof(template_coordinate_key_t))) == NULL) {
```

Comment:
This `realloc` is a bad idea because you're storing pointers into `keys` later on, which would become invalid if the data gets moved during reallocation.  Instead, you'd need to store a sequence of fixed buffers along with a list or array of where they start for use at demolition time.  Making them all the same size might be easiest as then you can just allocate a new one at the appropriate values of `k`, and the start points actually get stored in the `buf[k].u.key` pointers for the multiples of the array size you choose so there's no need to store them separately.